src/DBFTPlugin/ConsensusService.cs

```diff
@@ -423,17 +423,35 @@ private void OnPrepareRequestReceived(ExtensiblePayload payload, PrepareRequest
                 Log($"Timestamp incorrect: {message.Timestamp}", LogLevel.Warning);
                 return;
             }
+
             if (message.TransactionHashes.Any(p => NativeContract.Ledger.ContainsTransaction(context.Snapshot, p)))
             {
                 Log($"Invalid request: transaction already exists", LogLevel.Warning);
                 return;
             }
 
+            // Possible vulnerability of the random number, dishonest nodes cooperate to
+            // attck the system. 
+            byte[] nonce;
+            try
+            {
+                nonce = VRF.Verify(context.Validators[message.ValidatorIndex], message.VRFProof, message.PrevHash);
```

Comment:
We should add the current merkletree in `data`

Reply:
If we add the current Merkle tree in the `data`, the result would be mutable for the primary cause he can decide the Merkle tree.

Reply:
>  he can decide the Merkle tree.

But not the `PrevHash`, it makes the things more complicated

Reply:
Using the block hash is a normal and the most easy way to get a well known input. Currently, almost all projects that use VRF take block hash as input, either prehash or previous n blocks. But if u have a better input source, definitely it is a good thing.